pkg/kubelet/pod_workers.go

```diff
@@ -194,38 +194,51 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan UpdatePodOptions) {
 	}
 }
 
+func (p *podWorkers) getWorker(uid types.UID) chan UpdatePodOptions {
+	podUpdates, exists := p.podUpdates[uid]
+	if exists {
+		return podUpdates
+	}
+
+	// We need to have a buffer here, because checkForUpdates() method that
+	// puts an update into channel is called from the same goroutine where
+	// the channel is consumed. However, it is guaranteed that in such case
+	// the channel is empty, so buffer of size 1 is enough.
+	podUpdates = make(chan UpdatePodOptions, 1)
+	p.podUpdates[uid] = podUpdates
+
+	// Creating a new pod worker either means this is a new pod, or that the
+	// kubelet just restarted. In either case the kubelet is willing to believe
+	// the status of the pod for the first pod worker sync. See corresponding
+	// comment in syncPod.
+	go func() {
+		defer runtime.HandleCrash()
+		p.managePodLoop(podUpdates)
+	}()
+	return podUpdates
+}
+
 // Apply the new setting to the specified pod.
 // If the options provide an OnCompleteFunc, the function is invoked if the update is accepted.
 // Update requests are ignored if a kill pod request is pending.
 func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
 	pod := options.Pod
 	uid := pod.UID
-	var podUpdates chan UpdatePodOptions
-	var exists bool
 
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
-	if podUpdates, exists = p.podUpdates[uid]; !exists {
-		// We need to have a buffer here, because checkForUpdates() method that
-		// puts an update into channel is called from the same goroutine where
-		// the channel is consumed. However, it is guaranteed that in such case
-		// the channel is empty, so buffer of size 1 is enough.
-		podUpdates = make(chan UpdatePodOptions, 1)
-		p.podUpdates[uid] = podUpdates
-
-		// Creating a new pod worker either means this is a new pod, or that the
-		// kubelet just restarted. In either case the kubelet is willing to believe
-		// the status of the pod for the first pod worker sync. See corresponding
-		// comment in syncPod.
-		go func() {
-			defer runtime.HandleCrash()
-			p.managePodLoop(podUpdates)
-		}()
-	}
-	if !p.isWorking[pod.UID] {
-		p.isWorking[pod.UID] = true
-		podUpdates <- *options
-	} else {
+
+	working := p.isWorking[pod.UID]
+	if working == nil {
+		p.isWorking[pod.UID] = options
+		p.getWorker(uid) <- *options
+	} else if isShortenedGracePeriod(working, options) {
+		// The previous worker needs to wait until the pod is successfully deleted before executing the next one,
+		// we close the previous worker and start a new worker to perform the pod delete operation
+		p.removeWorker(pod.UID)
```

Comment:
just noting that this is right to call versus `ForgetWorker` as we hold the pod lock.